<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>CgXBP1_analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="CgXBP1_analysis_sep_2023_files/libs/clipboard/clipboard.min.js"></script>
<script src="CgXBP1_analysis_sep_2023_files/libs/quarto-html/quarto.js"></script>
<script src="CgXBP1_analysis_sep_2023_files/libs/quarto-html/popper.min.js"></script>
<script src="CgXBP1_analysis_sep_2023_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="CgXBP1_analysis_sep_2023_files/libs/quarto-html/anchor.min.js"></script>
<link href="CgXBP1_analysis_sep_2023_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="CgXBP1_analysis_sep_2023_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="CgXBP1_analysis_sep_2023_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="CgXBP1_analysis_sep_2023_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="CgXBP1_analysis_sep_2023_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">CgXBP1_analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">message=</span><span class="cn">FALSE</span>, <span class="at">warning=</span><span class="cn">FALSE</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/chip_analysis_functions.R"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnrichedHeatmap)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="polii-chip-seq-analysis" class="level1">
<h1>PolII ChIP-seq analysis</h1>
<section id="generate-normalised-gene-expression-values-commonly-known-as-fpkm-rpkm-tpm-etc.-for-polii-data." class="level2">
<h2 class="anchored" data-anchor-id="generate-normalised-gene-expression-values-commonly-known-as-fpkm-rpkm-tpm-etc.-for-polii-data.">Generate normalised gene expression values (Commonly known as FPKM, RPKM, TPM, etc. ) for PolII data.</h2>
<p>To generate the normalised expression values for each gene aligned reads were processed by <code>MACS2</code> (Zhang et al., 2008), and <code>BigWig</code> files were generated using <code>bedSort</code> and <code>bedGraphToBigWig</code> commands from <code>UCSC Kent utils</code> (Kent et al., 2010). <code>Samtools</code> (version 1.9) was used to index the resultant <code>BAM</code> file and check for alignment statistics. Elongating RNAPII occupancy was then measured by <code>normalizeToMatrix</code> function from the R package <code>EnrichedHeatmap</code>. Normalized values then saved to .rds files (see R script <code>generate_normalised_gene_expression_matrix_from_polII_chip_data.R</code>) which are then used for downstream analysis and visualizations.</p>
</section>
<section id="polii-occupancy-heatmap-wt" class="level2">
<h2 class="anchored" data-anchor-id="polii-occupancy-heatmap-wt">PolII occupancy heatmap (WT)</h2>
<div class="cell" data-hash="CgXBP1_analysis_sep_2023_cache/html/profile-hearmap-wt_90f0403f66ddb27f44925a407b73323d">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>gff_file <span class="ot">&lt;-</span> <span class="st">"../data/C_glabrata_CBS138_version_s02-m07-r06_features.gff"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>gff <span class="ot">&lt;-</span> GenomicFeatures<span class="sc">::</span><span class="fu">makeTxDbFromGFF</span>(gff_file, <span class="at">metadata =</span> T)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> GenomicFeatures<span class="sc">::</span><span class="fu">genes</span>(gff)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="do">## prepare signal data</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>bw_files_dir <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"../data/bw/WT"</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>bw_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(bw_files_dir, <span class="at">pattern =</span> <span class="st">"*set1.*bw"</span>, <span class="at">recursive =</span> T, <span class="at">full.names =</span> T)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bw_files) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"0.5h"</span>, <span class="st">"2h"</span>, <span class="st">"4h"</span>, <span class="st">"6h"</span>, <span class="st">"8h"</span>)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>bw_files <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">enframe</span>(bw_files)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="do">## generate normalised matrix in tidy way</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>wt_norm_mat <span class="ot">&lt;-</span> bw_files <span class="sc">%&gt;%</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">bw =</span> purrr<span class="sc">::</span><span class="fu">map</span>(value, <span class="cf">function</span>(ii) {</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>                rtracklayer<span class="sc">::</span><span class="fu">import</span>(ii)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>        })) <span class="sc">%&gt;%</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">norm_matrix =</span> purrr<span class="sc">::</span><span class="fu">map</span>(bw, <span class="cf">function</span>(ii) {</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>                nn <span class="ot">&lt;-</span> EnrichedHeatmap<span class="sc">::</span><span class="fu">normalizeToMatrix</span>(<span class="at">signal =</span> ii, </span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>                                                         <span class="at">target =</span> genes, </span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>                                                         <span class="at">value_column =</span> <span class="st">"score"</span>, </span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>                                                         <span class="at">smooth =</span> T, </span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>                                                         <span class="at">extend =</span> <span class="dv">200</span>)</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>                <span class="fu">return</span>(nn)</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>        }))</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="do">## plot hm list</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>wt_profile_heatmap <span class="ot">&lt;-</span> <span class="fu">get_enrichment_heatmap_list</span>(</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> wt_norm_mat<span class="sc">$</span>norm_matrix,</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> circlize<span class="sc">::</span><span class="fu">colorRamp2</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">50</span>, <span class="at">by =</span> <span class="dv">10</span>), <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"#fdd49e"</span>, <span class="st">"#fdbb84"</span>, <span class="st">"#fc8d59"</span>, <span class="st">"#e34a33"</span>, <span class="st">"#b30000"</span>)),</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>        <span class="at">names =</span> wt_norm_mat<span class="sc">$</span>name,</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">titles =</span> wt_norm_mat<span class="sc">$</span>name,</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis_name =</span> <span class="fu">c</span>(<span class="st">"-200"</span>, <span class="st">"TSS"</span>, <span class="st">"TES"</span>, <span class="st">"+200"</span>),</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis_name_rot =</span> <span class="dv">90</span> ,</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis_name_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">10</span>, <span class="at">fontface =</span> <span class="st">"bold"</span>),</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>        <span class="at">pos_line =</span> T,</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>        <span class="at">top_annotation =</span> <span class="fu">HeatmapAnnotation</span>(<span class="at">lines =</span> <span class="fu">anno_enriched</span>(<span class="at">axis_param =</span> <span class="fu">list</span>(<span class="at">side =</span> <span class="st">"right"</span>, <span class="at">facing =</span> <span class="st">"inside"</span>)))</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">draw</span>(wt_profile_heatmap)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/profile-hearmap-wt-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="prepare-data-for-normalised-gene-expression-heatmap" class="level2">
<h2 class="anchored" data-anchor-id="prepare-data-for-normalised-gene-expression-heatmap">Prepare data for normalised gene expression heatmap</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load FPKM matrix </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>wt_fpkm_mat <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(<span class="at">file =</span> <span class="st">"../data/rds/cg_polII_wt_thp1_sample_wise_fpkm.rds"</span>)   </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>wt_fpkm_mat <span class="ot">&lt;-</span> wt_fpkm_mat <span class="sc">%&gt;%</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>        tidyr<span class="sc">::</span><span class="fu">gather</span>(<span class="st">"Cond"</span>, <span class="st">"FPKM"</span>, <span class="sc">-</span>gene_name) <span class="sc">%&gt;%</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>        tidyr<span class="sc">::</span><span class="fu">separate</span>(<span class="at">col =</span> <span class="st">"Cond"</span>, <span class="at">sep =</span> <span class="st">"_"</span>, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"organism"</span>, <span class="st">"strain"</span>, <span class="st">"condition"</span>, <span class="st">"timepoint"</span>, <span class="st">"replicate"</span>))</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>xbp1_fpkm_mat <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(<span class="at">file =</span> <span class="st">"../data/rds/cg_polII_xbp1_thp1_sample_wise_fpkm.rds"</span>)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>xbp1_fpkm_mat <span class="ot">&lt;-</span> xbp1_fpkm_mat <span class="sc">%&gt;%</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>        tidyr<span class="sc">::</span><span class="fu">gather</span>(<span class="st">"Cond"</span>, <span class="st">"FPKM"</span>, <span class="sc">-</span>gene_name) <span class="sc">%&gt;%</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>        tidyr<span class="sc">::</span><span class="fu">separate</span>(<span class="at">col =</span> <span class="st">"Cond"</span>, <span class="at">sep =</span> <span class="st">"_"</span>, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"organism"</span>, <span class="st">"strain"</span>, <span class="st">"condition"</span>, <span class="st">"timepoint"</span>, <span class="st">"replicate"</span>))</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="do">## combine datasets</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>cg_polII_data <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(wt_fpkm_mat, xbp1_fpkm_mat) <span class="sc">%&gt;%</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">group_by</span>(gene_name, strain, condition, replicate) <span class="sc">%&gt;%</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">stdev =</span> <span class="fu">sd</span>(FPKM)) <span class="do">## calculate sd within group across all timepoints . Make sure that number of elem in each group must be equal to time points.</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(wt_fpkm_mat, xbp1_fpkm_mat) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">group_by</span>(gene_name, strain, condition, replicate) <span class="sc">%&gt;%</span> </span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">tally</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">arrange</span>(<span class="fu">desc</span>(n))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 21,244 × 5
# Groups:   gene_name, strain, condition [10,622]
   gene_name    strain condition replicate     n
   &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;     &lt;int&gt;
 1 CAGL0A00105g wt     thp1      set1          5
 2 CAGL0A00105g wt     thp1      set2          5
 3 CAGL0A00105g xbp1   thp1      set1          5
 4 CAGL0A00105g xbp1   thp1      set2          5
 5 CAGL0A00110g wt     thp1      set1          5
 6 CAGL0A00110g wt     thp1      set2          5
 7 CAGL0A00110g xbp1   thp1      set1          5
 8 CAGL0A00110g xbp1   thp1      set2          5
 9 CAGL0A00116g wt     thp1      set1          5
10 CAGL0A00116g wt     thp1      set2          5
# ℹ 21,234 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## cross check the numbers (e.g number of unique genes, organism , strain, conditions etc.)</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>cg_polII_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">summarise_all</span>(dplyr<span class="sc">::</span>n_distinct)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 8
  gene_name organism strain condition timepoint replicate   FPKM stdev
      &lt;int&gt;    &lt;int&gt;  &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;  &lt;int&gt; &lt;int&gt;
1      5311        1      2         1         5         2 106090 21237</code></pre>
</div>
</div>
</section>
<section id="correlation-of-polii-replicates-using-fpkm-values-wt" class="level2">
<h2 class="anchored" data-anchor-id="correlation-of-polii-replicates-using-fpkm-values-wt">Correlation of PolII replicates using FPKM values (WT)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>wt_data_for_corr <span class="ot">&lt;-</span> cg_polII_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(strain <span class="sc">==</span> <span class="st">"wt"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">id_cols  =</span> <span class="fu">c</span>(gene_name, timepoint), </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">names_from =</span> replicate, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">values_from =</span> FPKM) </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>plot_wt_cor <span class="ot">&lt;-</span> wt_data_for_corr <span class="sc">%&gt;%</span> </span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">set1 =</span> <span class="fu">log2</span>(set1<span class="sc">+</span><span class="dv">1</span>),<span class="at">set2 =</span> <span class="fu">log2</span>(set2<span class="sc">+</span><span class="dv">1</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  ggpubr<span class="sc">::</span><span class="fu">ggscatter</span>(<span class="at">x =</span> <span class="st">"set1"</span>,<span class="at">y =</span> <span class="st">"set2"</span>, <span class="at">facet.by =</span> <span class="st">"timepoint"</span>, <span class="at">cor.coef =</span> T ,<span class="at">cor.coeff.args =</span> <span class="fu">list</span>(<span class="at">col =</span> <span class="st">"red"</span>), <span class="at">nrow =</span> <span class="dv">2</span> ) <span class="sc">+</span> </span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"WT"</span>) <span class="sc">+</span> </span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">5</span>,<span class="dv">10</span>)) <span class="sc">+</span> <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">5</span>,<span class="dv">10</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">10</span>))</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>plot_wt_cor</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-1-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="correlation-of-polii-replicates-using-fpkm-values-xbp1" class="level2">
<h2 class="anchored" data-anchor-id="correlation-of-polii-replicates-using-fpkm-values-xbp1">Correlation of PolII replicates using FPKM values (xbp1)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>xbp1_data_for_corr <span class="ot">&lt;-</span> cg_polII_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(strain <span class="sc">==</span> <span class="st">"xbp1"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">id_cols  =</span> <span class="fu">c</span>(gene_name, timepoint), </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">names_from =</span> replicate, </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">values_from =</span> FPKM) </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>plot_xbp1_cor <span class="ot">&lt;-</span> xbp1_data_for_corr <span class="sc">%&gt;%</span> </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">set1 =</span> <span class="fu">log2</span>(set1<span class="sc">+</span><span class="dv">1</span>),<span class="at">set2 =</span> <span class="fu">log2</span>(set2<span class="sc">+</span><span class="dv">1</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  ggpubr<span class="sc">::</span><span class="fu">ggscatter</span>(<span class="at">x =</span> <span class="st">"set1"</span>,<span class="at">y =</span> <span class="st">"set2"</span>, <span class="at">facet.by =</span> <span class="st">"timepoint"</span>, <span class="at">cor.coef =</span> T ,<span class="at">cor.coeff.args =</span> <span class="fu">list</span>(<span class="at">col =</span> <span class="st">"red"</span>), <span class="at">nrow =</span> <span class="dv">2</span> ) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"xbp1∆"</span>)  <span class="sc">+</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">5</span>,<span class="dv">10</span>)) <span class="sc">+</span> <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">5</span>,<span class="dv">10</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">10</span>))</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>plot_xbp1_cor</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="plot-sd-distributions-for-each-replicate" class="level2">
<h2 class="anchored" data-anchor-id="plot-sd-distributions-for-each-replicate">Plot SD distributions for each replicate</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>sd_disturb_plot <span class="ot">&lt;-</span> cg_polII_data <span class="sc">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">slice</span>(1L) <span class="sc">%&gt;%</span> <span class="do">## get one elem of each group</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x =</span> stdev), <span class="at">binwidth =</span> <span class="dv">1</span>, <span class="at">fill =</span> <span class="st">"cyan3"</span>, <span class="at">col =</span> <span class="st">"black"</span>) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">50</span>) <span class="sc">+</span> <span class="fu">facet_wrap</span>(replicate <span class="sc">~</span> strain) <span class="sc">+</span> <span class="fu">theme_bw</span>()</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>sd_disturb_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/sd-histograms-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="count-number-of-genes-at-several-sd-cutoffs-wt" class="level2">
<h2 class="anchored" data-anchor-id="count-number-of-genes-at-several-sd-cutoffs-wt">Count number of genes at several SD cutoffs (WT)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>gp <span class="ot">&lt;-</span> <span class="fu">counts_by_sd_cutoffs</span>(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> wt_fpkm_mat <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                tidyr<span class="sc">::</span><span class="fu">spread</span>(timepoint, FPKM) <span class="sc">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                dplyr<span class="sc">::</span><span class="fu">filter</span>(replicate <span class="sc">==</span> <span class="st">"set1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>                dplyr<span class="sc">::</span><span class="fu">select_if</span>(is.double),</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">y =</span> <span class="fu">paste</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">8</span>), <span class="st">"h"</span>, <span class="at">sep =</span> <span class="st">""</span>),</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">sds_cutoffs =</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="fl">0.5</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>gp <span class="sc">+</span> <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/count-genes-by-sd-wt-1.png" class="img-fluid" width="384"></p>
</div>
</div>
</section>
<section id="polii-gene-expression-heatmap-wt" class="level2">
<h2 class="anchored" data-anchor-id="polii-gene-expression-heatmap-wt">PolII gene expression heatmap (WT)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="do">####  zscore heatmap plot</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="do">## @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>sd_cut_off <span class="ot">&lt;-</span> <span class="fl">2.25</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>strain_type <span class="ot">&lt;-</span> <span class="st">"wt"</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>rep <span class="ot">&lt;-</span> <span class="st">"set1"</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>fpkm_mat <span class="ot">&lt;-</span> cg_polII_data <span class="sc">%&gt;%</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">filter</span>(stdev <span class="sc">&gt;=</span> sd_cut_off, replicate <span class="sc">==</span> rep) <span class="sc">%&gt;%</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(gene_name, strain, timepoint, FPKM) <span class="sc">%&gt;%</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>        tidyr<span class="sc">::</span><span class="fu">unite</span>(<span class="at">col =</span> <span class="st">"ident"</span>, <span class="at">sep =</span> <span class="st">"_"</span>, strain, timepoint, <span class="at">remove =</span> F) <span class="sc">%&gt;%</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>timepoint) <span class="sc">%&gt;%</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">group_by</span>(gene_name)</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>zscore_mats <span class="ot">&lt;-</span> fpkm_mat  <span class="sc">%&gt;%</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">group_by</span>( strain, gene_name) <span class="sc">%&gt;%</span> </span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">zscore =</span> <span class="fu">scale</span>(FPKM)) <span class="sc">%&gt;%</span> </span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>FPKM) <span class="sc">%&gt;%</span> </span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>        tidyr<span class="sc">::</span><span class="fu">spread</span>(<span class="at">key =</span> ident, <span class="at">value =</span> zscore) <span class="sc">%&gt;%</span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">group_by</span>(strain) <span class="sc">%&gt;%</span> </span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">group_split</span>() <span class="sc">%&gt;%</span></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>        purrr<span class="sc">::</span><span class="fu">map</span>(<span class="sc">~</span> ..<span class="dv">1</span> <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>strain) <span class="sc">%&gt;%</span> TidyWrappers<span class="sc">::</span><span class="fu">tbl_remove_vars_NA_any</span>())</span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a><span class="co"># plot WT heatmap</span></span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a>hm_data <span class="ot">&lt;-</span> zscore_mats[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a>        tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"gene_name"</span>)</span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a>seed <span class="ot">&lt;-</span> <span class="dv">7173</span></span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(seed)</span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a>wt_km <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(hm_data, <span class="at">centers =</span> <span class="dv">8</span>)</span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(wt_km<span class="sc">$</span>cluster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  1   2   3   4   5   6   7   8 
166 149 164 219  85 217 151  77 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>wt_cl_order <span class="ot">&lt;-</span> <span class="fu">factor</span>(wt_km<span class="sc">$</span>cluster, <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">6</span>,<span class="dv">4</span>,<span class="dv">7</span>,<span class="dv">8</span>,<span class="dv">5</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>        forcats<span class="sc">::</span><span class="fu">fct_relabel</span>(<span class="sc">~</span><span class="fu">paste</span>(<span class="st">"clust"</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>, <span class="at">sep =</span> <span class="st">"-"</span>))</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>cg_wt_heatmap <span class="ot">=</span> ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(hm_data, <span class="at">show_row_dend =</span> F, </span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">show_row_names =</span> F, </span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">cluster_columns =</span> F, </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">show_column_dend =</span> F, </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">row_split =</span> wt_cl_order, </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>                        <span class="at">name =</span> <span class="st">"Z-Score"</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>                        <span class="at">cluster_row_slices =</span> F)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>cg_wt_heatmap <span class="ot">&lt;-</span> ComplexHeatmap<span class="sc">::</span><span class="fu">draw</span>(cg_wt_heatmap)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/zscore-wt-heatmap-1.png" class="img-fluid" width="288"></p>
</div>
</div>
</section>
<section id="polii-gene-expression-heatmap-xbp1" class="level2">
<h2 class="anchored" data-anchor-id="polii-gene-expression-heatmap-xbp1">PolII gene expression heatmap (xbp1)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>hm_data <span class="ot">&lt;-</span> zscore_mats[[<span class="dv">2</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>        tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"gene_name"</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>seed <span class="ot">&lt;-</span> <span class="dv">7173</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(seed)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>km <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(hm_data, <span class="at">centers =</span> <span class="dv">7</span>)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(km<span class="sc">$</span>cluster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  1   2   3   4   5   6   7 
400 329 131 506 229 249 302 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>cl_order <span class="ot">&lt;-</span> <span class="fu">factor</span>(km<span class="sc">$</span>cluster, <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">1</span>,<span class="dv">5</span>,<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">7</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>        forcats<span class="sc">::</span><span class="fu">fct_relabel</span>(<span class="sc">~</span><span class="fu">paste</span>(<span class="st">"clust"</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>, <span class="at">sep =</span> <span class="st">"-"</span>))</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>cg_xbp1_heatmap <span class="ot">=</span> ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(hm_data, <span class="at">show_row_dend =</span> F, </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">show_row_names =</span> F, </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">cluster_columns =</span> F, </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">show_column_dend =</span> F, </span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">row_split =</span> cl_order, </span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>                        <span class="at">name =</span> <span class="st">"Z-Score"</span>,</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>                        <span class="at">cluster_row_slices =</span> F)</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>cg_xbp1_heatmap</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/zscore-xbp1-heatmap-1.png" class="img-fluid" width="288"></p>
</div>
</div>
</section>
<section id="plot-transcription-factor-wt-temporally" class="level2">
<h2 class="anchored" data-anchor-id="plot-transcription-factor-wt-temporally">Plot transcription factor (WT) Temporally</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>cg_tf <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_lines</span>(<span class="st">"../data/Cg_polII_wt_temporal_TFs.txt"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>tf_hm_data <span class="ot">&lt;-</span> wt_cl_order[<span class="fu">names</span>(wt_cl_order) <span class="sc">%in%</span> cg_tf] <span class="sc">%&gt;%</span> </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">enframe</span>(<span class="at">name =</span> <span class="st">"gene_name"</span>, <span class="at">value =</span> <span class="st">"clust"</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(zscore_mats[[<span class="dv">1</span>]] , <span class="at">by =</span> <span class="st">"gene_name"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>        tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"gene_name"</span>) </span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>cg_tf_heatmap <span class="ot">=</span> ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(tf_hm_data[,<span class="sc">-</span><span class="dv">1</span>], <span class="at">show_row_dend =</span> F, </span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>                        <span class="at">show_row_names =</span> T, </span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>                        <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">cex =</span> <span class="fl">0.7</span>),</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>                        <span class="at">cluster_columns =</span> F, </span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>                        <span class="at">show_column_dend =</span> F,</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>                        <span class="at">column_title_gp =</span> <span class="fu">gpar</span>(<span class="at">cex =</span> <span class="fl">0.8</span>),</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>                        <span class="at">row_split =</span> tf_hm_data<span class="sc">$</span>clust, </span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>                        <span class="at">name =</span> <span class="st">"Z-Score"</span>,</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>                        <span class="at">column_title =</span> <span class="st">"Temporally Expressed Transcription Factors"</span>,</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>                        <span class="at">cluster_row_slices =</span> F)</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>cg_tf_heatmap</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="384"></p>
</div>
</div>
</section>
<section id="pathway-specific-heatmap" class="level2">
<h2 class="anchored" data-anchor-id="pathway-specific-heatmap">Pathway specific heatmap</h2>
<section id="prepare-data" class="level3">
<h3 class="anchored" data-anchor-id="prepare-data">Prepare data</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>excel_sheets <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"DNA_repair"</span>,<span class="st">"iron_homeostasis"</span>,<span class="st">"oxidative_stress"</span>,<span class="st">"gluconeogenesis"</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>              ,<span class="st">"DNA_D_cc_arrest_in_resp_N_starv"</span>,<span class="st">"TCA_cycle"</span>,<span class="st">"autophagy"</span>,<span class="st">"Aminoacid_biosynthesis"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>gene_sets <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(excel_sheets, <span class="sc">~</span>readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">"../data/pathway_genes.xlsx"</span>, <span class="at">sheet =</span> ..<span class="dv">1</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(gene_name, gene_symbol, cluster))</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gene_sets) <span class="ot">&lt;-</span> excel_sheets</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># prepare data</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>pathway_genes_hm_data <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(gene_sets, <span class="sc">~</span> ..<span class="dv">1</span> <span class="sc">%&gt;%</span> </span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(cg_polII_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(strain <span class="sc">==</span> <span class="st">"wt"</span><span class="sc">&amp;</span> replicate <span class="sc">==</span> <span class="st">"set1"</span>), <span class="at">by =</span> <span class="st">"gene_name"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>( strain, gene_name) <span class="sc">%&gt;%</span> </span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">zscore =</span> <span class="fu">scale</span>(FPKM)) <span class="sc">%&gt;%</span> </span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>FPKM) <span class="sc">%&gt;%</span> </span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">"timepoint"</span>, <span class="at">values_from  =</span> <span class="st">"zscore"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>() <span class="sc">%&gt;%</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">c</span>(gene_symbol, cluster, dplyr<span class="sc">::</span><span class="fu">ends_with</span>(<span class="st">"h"</span>))) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"gene_symbol"</span>) </span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a><span class="co"># plot heatmap</span></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>pathway_heatmaps <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(<span class="fu">names</span>(pathway_genes_hm_data) ,<span class="sc">~</span> ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(<span class="at">matrix =</span> pathway_genes_hm_data[[..<span class="dv">1</span>]][,<span class="sc">-</span><span class="dv">1</span>], <span class="at">row_split =</span> pathway_genes_hm_data[[..<span class="dv">1</span>]]<span class="sc">$</span>cluster, <span class="at">cluster_row_slices =</span> F, <span class="at">cluster_columns =</span> F, <span class="at">show_row_dend =</span> F, <span class="at">column_title =</span> ..<span class="dv">1</span>, <span class="at">name =</span> <span class="st">"Z-Score"</span>)</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a><span class="co"># plot </span></span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>purrr<span class="sc">::</span><span class="fu">walk</span>(pathway_heatmaps, <span class="sc">~</span>ComplexHeatmap<span class="sc">::</span><span class="fu">draw</span>(..<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="480"></p>
</div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-4-2.png" class="img-fluid" width="480"></p>
</div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-4-3.png" class="img-fluid" width="480"></p>
</div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-4-4.png" class="img-fluid" width="480"></p>
</div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-4-5.png" class="img-fluid" width="480"></p>
</div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-4-6.png" class="img-fluid" width="480"></p>
</div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-4-7.png" class="img-fluid" width="480"></p>
</div>
<div class="cell-output-display">
<p><img src="CgXBP1_analysis_sep_2023_files/figure-html/unnamed-chunk-4-8.png" class="img-fluid" width="480"></p>
</div>
</div>
</section>
</section>
<section id="section" class="level2">
<h2 class="anchored" data-anchor-id="section"><!--# Prepare data for pathway specific heatmap --></h2>
<!--#  -->
</section>
<section id="section-1" class="level2">
<h2 class="anchored" data-anchor-id="section-1"><!--# Pathway specific heatmap: WT - TCA cycle --></h2>
<!--#  -->
</section>
<section id="section-2" class="level2">
<h2 class="anchored" data-anchor-id="section-2"><!--# Pathway specific heatmap: xbp1 - TCA cycle --></h2>
<!--#  -->
</section>
<section id="section-3" class="level2">
<h2 class="anchored" data-anchor-id="section-3"><!--# Pathway specific heatmap: WT - Amino acid biosynthesis --></h2>
<!--#  -->
</section>
<section id="section-4" class="level2">
<h2 class="anchored" data-anchor-id="section-4"><!--# Pathway specific heatmap: xbp1 - Amino acid biosynthesis --></h2>
<!--#  -->
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>